src/Http/Routing/src/DefaultLinkParser.cs

```diff
@@ -0,0 +1,236 @@
+                {
+                    constraints.Add(kvp.Key, constraintsForParameter);
+                }
+            }
+
+            var matcher = new RoutePatternMatcher(endpoint.RoutePattern, new RouteValueDictionary(endpoint.RoutePattern.Defaults));
+            return new MatcherState(matcher, constraints);
+        }
+
+        // Internal for testing
+        internal MatcherState GetMatcherState(RouteEndpoint endpoint) => _matcherCache.EnsureInitialized().GetOrAdd(endpoint, _createMatcher);
+
+        // Internal for testing
+        internal bool TryParse(RouteEndpoint endpoint, PathString path, out RouteValueDictionary values)
+        {
+            var (matcher, constraints) = GetMatcherState(endpoint);
+
+            values = new RouteValueDictionary();
+            if (!matcher.TryMatch(path, values))
+            {
+                values = null;
+                return false;
+            }
+
+            foreach (var kvp in constraints)
+            {
+                for (var i = 0; i < kvp.Value.Count; i++)
+                {
+                    var constraint = kvp.Value[i];
+                    if (!constraint.Match(httpContext: null, NullRouter.Instance, kvp.Key, values, RouteDirection.IncomingRequest))
+                    {
+                        values = null;
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        public void Dispose()
+        {
+            _matcherCache.Dispose();
+        }
+
+        // internal for testing
+        internal readonly struct MatcherState
+        {
+            public readonly RoutePatternMatcher Matcher;
+            public readonly Dictionary<string, List<IRouteConstraint>> Constraints;
+
+            public MatcherState(RoutePatternMatcher matcher, Dictionary<string, List<IRouteConstraint>> constraints)
+            {
+                Matcher = matcher;
+                Constraints = constraints;
+            }
+
+            public void Deconstruct(out RoutePatternMatcher matcher, out Dictionary<string, List<IRouteConstraint>> constraints)
+            {
+                matcher = Matcher;
+                constraints = Constraints;
+            }
+        }
+
+        private static class Log
+        {
+            public static class EventIds
+            {
+                public static readonly EventId EndpointsFound = new EventId(100, "EndpointsFound");
+                public static readonly EventId EndpointsNotFound = new EventId(101, "EndpointsNotFound");
+
+                public static readonly EventId PathParsingSucceeded = new EventId(102, "PathParsingSucceeded");
+                public static readonly EventId PathParsingFailed = new EventId(103, "PathParsingFailed");
+            }
+
+            private static readonly Action<ILogger, IEnumerable<string>, object, Exception> _endpointsFound = LoggerMessage.Define<IEnumerable<string>, object>(
```

Comment:
I didn't know logging supported enumerable. Could you add a unit test that checks that the log message looks like what you expected when there are multiple matches.

Alternatively I think it is fine to just have a count in the log message.

Reply:
I did know that  😆